File: go/002AddTwoNumbers/002AddTwoNumbers.go
package AddTwoNumbers

/*
You are given two linked lists representing two non-negative numbers. 
The digits are stored in reverse order and each of their nodes contain a single digit. 
Add the two numbers and return it as a linked list.

Input: (2 -> 4 -> 3) + (5 -> 6 -> 4)
Output: 7 -> 0 -> 8
*/

//Definition for singly-linked list.
type ListNode struct {
	Val int
    Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	dummy := &ListNode{0, nil}
	var q,l *ListNode

	c := 0
	for l1!=nil && l2!=nil{
		p := &ListNode{l1.Val + l2.Val + c, nil}
		if p.Val >=10 {
			c = 1
			p.Val -= 10
		}else{
			c = 0
		}
		if dummy.Next == nil{
			dummy.Next = p
		}else{
			q.Next = p
		}
		q = p
		l1 = l1.Next
		l2 = l2.Next
	}

	if l1!= nil{
		l = l1
	}else if l2!=nil{
		l = l2
	}else{
		l = nil
	}

	for l!=nil{
		p := &ListNode{l.Val + c, nil}
		if p.Val >=10 {
			c = 1
			p.Val -= 10
		}else{
			c = 0
		}
		if dummy.Next == nil{
			dummy.Next = p
		}else{
			q.Next = p
		}
		q = p	
		l = l.Next
	}

	if c!=0 {
		p := &ListNode{c, nil}
		if dummy.Next == nil{
			dummy.Next = p
		}else{
			q.Next = p
		}
	}

	return dummy.Next;
}

File: src/002AddTwoNumbers/002AddTwoNumbers.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn add_two_numbers(l1: Option<Box<ListNode>>, l2: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut dummy = ListNode::new(0);
        let mut l = &mut dummy;
        let mut l1 = l1;
        let mut l2 = l2;
        let mut c = 0;
        while let (Some(p1), Some(p2)) = (l1.as_mut(), l2.as_mut()) {
            let d = p1.val + p2.val + c;
            c = d/10;
            l.next = Some(Box::new(ListNode::new(d%10)));
            l = l.next.as_mut().unwrap();
            l1 = p1.next.take();
            l2 = p2.next.take();
        }
        let mut l3 = if l1.is_some() {
            l1
        }else{
            l2
        };
        while let Some(p3) = l3.as_mut() {
            let d = p3.val + c;
            c = d/10;
            l.next = Some(Box::new(ListNode::new(d%10)));
            l = l.next.as_mut().unwrap();
            l3 = p3.next.take();
        }
        if c != 0 {
            l.next = Some(Box::new(ListNode::new(c)));
        }
        
        return dummy.next;
    }
}

// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution2 {
    pub fn add_two_numbers(l1: Option<Box<ListNode>>, l2: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut dummy = ListNode::new(0);
        let mut l = &mut dummy;
        let mut l1 = l1;
        let mut l2 = l2;
        let mut c = 0;
        while l1.is_some() || l2.is_some() {
            let v1 = if let Some(p1) = l1.as_mut() {
                let v = p1.val;
                l1 = p1.next.take();
                v
            }else{
                0
            };
            let v2 = if let Some(p2) = l2.as_mut() {
                let v = p2.val;
                l2 = p2.next.take();
                v
            }else{
                0
            };
            let d = v1 + v2 + c;
            c = d/10;
            l.next = Some(Box::new(ListNode::new(d%10)));
            l = l.next.as_mut().unwrap();
        }
        if c != 0 {
            l.next = Some(Box::new(ListNode::new(c)));
        }
        
        return dummy.next;
    }
}
